### ISSUE TITLE
Move from Flow to TypeScript

### ISSUE THREAD

**Issue by 100terres** [react-forked/dnd#8]
### Description

Related to: https://github.com/atlassian/react-beautiful-dnd/issues/982

Move our codebase to typescript while keeping flowtyped support.

**Comment by finnatinform** [comment]
Hey @100terres, great to see you effort on this fork! 😻 How can we support you to get this done?

**Comment by 100terres** [comment]
Hi @finnatinform, I’m currently working on migrating the test suit to `TypeScript` (https://github.com/react-forked/dnd/issues/15) and I’ll follow with the migration of the source code (https://github.com/react-forked/dnd/issues/10). I should be able to have a major part of the work done in a couple of days. I think what would be appreciated is to have other people review the code. I’ll do my best to make the review process as easy as it can be, but since it will affect almost all files it might be cumbersome task to do :sweat_smile:. I’ll also need help to keep `Flow` support (type declaration), since I’m not a `Flow` developer/user. After that we’ll have a clean slate to really build on top of and start tackling issues opened on [react-beautiful-dnd](https://github.com/atlassian/react-beautiful-dnd/issues). If you have any suggestions feel free to comment and/or open issues! :slightly_smiling_face:

**Comment by JamesGelok** [comment]
We could use flowgen to help make flow types from our typescript types

https://github.com/joarwilk/flowgen 

(I'm not a flow developer either 😅)

**Comment by 100terres** [comment]
@JamesGelok Thank you for the suggestion! It's what I've used! :slightly_smiling_face: You can take a look at where I'm currently at :point_right:  [here](https://github.com/react-forked/dnd/tree/issue-15-move-our-tests-to-typescript). :slightly_smiling_face:

**Comment by cmacdonnacha** [comment]
Hey @100terres , just wondering if there's anything we can help with here?

**Comment by 100terres** [comment]
Hi @cmacdonnacha thanks for pinging me! I was a bit busy at work in the last couple weeks, but I recently managed to migrate all the files in the `src` folder to TypeScript. I'll open a couple of PRs during the weekend and I'll add you as a reviewer :slightly_smiling_face:

After that I've already started the migrations of the `test` folder to TypeScript, but there is still work to do to properly finished migration to TypeScript. If you want, once the migration of the `src` folder is merged, I could leave it to you to finished the migration of the `test` folder? :slightly_smiling_face:

**Comment by 100terres** [comment]
Hi there! Just to give you all feedback! Sorry, to have left this work unfinished! I'll push to finish this and really take the time to maintain this repo :slightly_smiling_face:

Sorry again and I won't let you down this time :slightly_smiling_face:

**Comment by 100terres** [comment]
Here's an update of where we currently are:

- Migrating `src/` from Flow to TypeScript :heavy_check_mark:
- Fix all TypeScript errors from the `src/` :heavy_check_mark: 
- Migrating `test/` from Flow to TypeScript :heavy_check_mark: 
- Fix all TypeScript errors from the `test/` :no_entry: We still have more than 549 errors to fix, but the migration is going smootly :slightly_smiling_face: 

I'll try to give you an other update in the upcoming days!

**Comment by 100terres** [comment]
We are now at 487 errors :slightly_smiling_face:

**Comment by 100terres** [comment]
We are now at 395 TypeScript/typings errors.  Just to mention the tests themselves are passing :slightly_smiling_face:

**Comment by 100terres** [comment]
We are now at 138 TypeScript errors! Almost done! :slightly_smiling_face:

**Comment by 100terres** [comment]
Will be done soon! Only 59 errors left :slightly_smiling_face:

**Comment by 100terres** [comment]
No more TypeScript errors :heavy_check_mark: 

Next step I'll do some clean-up :broom:

**Comment by 100terres** [comment]
I'll be able to do release `v14.0.0-alpha.5` soon :slightly_smiling_face:

**Comment by 100terres** [comment]
The only thing missing here is the support of flow declaration:

- #132 

But for now it will be a nice to have. If anyone would like to tackle this issue and open a PR I'll gladly review it and merge it :slightly_smiling_face:

This conclude the migration from flow to TypeScript!

**Comment by 100terres** [comment]
v14.0.0 released :slightly_smiling_face:
